### PR TITLE
Allow solutions and tests in the same file

### LINKED ISSUE #776: Fix D language
The D language is unusual in that the tests _and_ implementation are in the same file. When submitting a valid D solution, the test runner fails though. After doing some digging, it looks like the test runner does not get the submitted code, but instead the original stub file (which is also the test file). This sort of makes sense, as we don't allow the students to change the test, but in this case that is problematic.

### PR DESCRIPTION
**Pull request by iHiD**
Fixes https://github.com/exercism/website/issues/776

### CODE REVIEW

File: app/models/git/exercise.rb [hunk 1]
```diff
@@ -33,6 +33,17 @@ def synced_git_sha
       commit.oid
     end
 
+    def valid_submission_filepath?(filepath)
+      return false if filepath.match?(%r{[^a-zA-Z0-9_./-]})
+      return false if filepath.starts_with?(".meta")
```
Comment (ErikSchierboom) [inline app/models/git/exercise.rb:38]: Do we want to exclude `.docs` files too?

File: app/models/git/exercise.rb [hunk 2]
```diff
@@ -33,6 +33,17 @@ def synced_git_sha
...
+
+      # Some languages have solutions and tests in the same file so
+      # we need to protect for that here.
+      return false if test_filepaths.include?(filepath) && !solution_filepaths.include?(filepath)
+
+      true
```
Comment (ErikSchierboom) [inline app/models/git/exercise.rb:45]: The if statement seems to cover the regular case whereas the comment describes the exceptional case

**Review by ErikSchierboom — APPROVED** [review]
I just tested this locally on the D track, and it works!

File: app/models/git/exercise.rb (hunk 1 above)
Comment (iHiD) [reply app/models/git/exercise.rb:38]: Sure :)

File: app/models/git/exercise.rb (hunk 2 above)
Comment (iHiD) [reply app/models/git/exercise.rb:45]: The bit after `&&` is the bit that is referred to by the comment. I'll make it clearer.

**Pull request merged**